
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="classification.html">
      
      
        <link rel="next" href="visualization.html">
      
      
        
      
      
      <link rel="icon" href="assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.7.2">
    
    
      
        <title>Barcode calling - IsoQuant</title>
      
    
    
      <link rel="stylesheet" href="assets/stylesheets/main.484c7ddc.min.css">
      
        
        <link rel="stylesheet" href="assets/stylesheets/palette.ab4e12ef.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL(".",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="white" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#barcode-calling" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="index.html" title="IsoQuant" class="md-header__button md-logo" aria-label="IsoQuant" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            IsoQuant
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Barcode calling
            
          </span>
        </div>
      </div>
    </div>
    
      
    
    
    
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="index.html" title="IsoQuant" class="md-nav__button md-logo" aria-label="IsoQuant" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    IsoQuant
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="index.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Home
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="getting-started.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Quick start
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="installation.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Installation
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="data.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Supported data
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" >
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Running IsoQuant
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            
  
    Running IsoQuant
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="input.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Ways of providing input data
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="cmd.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Command line options
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="examples.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Examples
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="single_cell.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Single-cell analysis
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6" >
        
          
          <label class="md-nav__link" for="__nav_6" id="__nav_6_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    IsoQuant output
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6">
            <span class="md-nav__icon md-icon"></span>
            
  
    IsoQuant output
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="output.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Output files
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="formats.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    File formats
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="classification.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Assignment classification
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    
  
    Barcode calling
  

    
  </span>
  
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="barcode_calling.html" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    
  
    Barcode calling
  

    
  </span>
  
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#how-it-works" class="md-nav__link">
    <span class="md-ellipsis">
      
        How it works
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#integration-with-isoquant" class="md-nav__link">
    <span class="md-ellipsis">
      
        Integration with IsoQuant
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#standalone-usage" class="md-nav__link">
    <span class="md-ellipsis">
      
        Standalone usage
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Standalone usage">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#command-line-options" class="md-nav__link">
    <span class="md-ellipsis">
      
        Command line options
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#example-commands" class="md-nav__link">
    <span class="md-ellipsis">
      
        Example commands
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#output" class="md-nav__link">
    <span class="md-ellipsis">
      
        Output
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Output">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#common-columns-all-modes" class="md-nav__link">
    <span class="md-ellipsis">
      
        Common columns (all modes)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#additional-columns-by-mode" class="md-nav__link">
    <span class="md-ellipsis">
      
        Additional columns by mode
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#statistics-file" class="md-nav__link">
    <span class="md-ellipsis">
      
        Statistics file
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#supported-platforms" class="md-nav__link">
    <span class="md-ellipsis">
      
        Supported platforms
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Supported platforms">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#10x-genomics-tenx_v3-visium_5prime" class="md-nav__link">
    <span class="md-ellipsis">
      
        10x Genomics (tenX_v3, visium_5prime)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#visium-hd-visium_hd" class="md-nav__link">
    <span class="md-ellipsis">
      
        Visium HD (visium_hd)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#curio-bioscience-curio" class="md-nav__link">
    <span class="md-ellipsis">
      
        Curio Bioscience (curio)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#stereo-seq-stereoseq-stereoseq_nosplit" class="md-nav__link">
    <span class="md-ellipsis">
      
        Stereo-seq (stereoseq, stereoseq_nosplit)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#custom-platform-custom_sc" class="md-nav__link">
    <span class="md-ellipsis">
      
        Custom platform (custom_sc)
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="visualization.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Visualization
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="citation.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Citation
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="feedback.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Feedback
  

    
  </span>
  
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#how-it-works" class="md-nav__link">
    <span class="md-ellipsis">
      
        How it works
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#integration-with-isoquant" class="md-nav__link">
    <span class="md-ellipsis">
      
        Integration with IsoQuant
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#standalone-usage" class="md-nav__link">
    <span class="md-ellipsis">
      
        Standalone usage
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Standalone usage">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#command-line-options" class="md-nav__link">
    <span class="md-ellipsis">
      
        Command line options
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#example-commands" class="md-nav__link">
    <span class="md-ellipsis">
      
        Example commands
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#output" class="md-nav__link">
    <span class="md-ellipsis">
      
        Output
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Output">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#common-columns-all-modes" class="md-nav__link">
    <span class="md-ellipsis">
      
        Common columns (all modes)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#additional-columns-by-mode" class="md-nav__link">
    <span class="md-ellipsis">
      
        Additional columns by mode
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#statistics-file" class="md-nav__link">
    <span class="md-ellipsis">
      
        Statistics file
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#supported-platforms" class="md-nav__link">
    <span class="md-ellipsis">
      
        Supported platforms
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Supported platforms">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#10x-genomics-tenx_v3-visium_5prime" class="md-nav__link">
    <span class="md-ellipsis">
      
        10x Genomics (tenX_v3, visium_5prime)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#visium-hd-visium_hd" class="md-nav__link">
    <span class="md-ellipsis">
      
        Visium HD (visium_hd)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#curio-bioscience-curio" class="md-nav__link">
    <span class="md-ellipsis">
      
        Curio Bioscience (curio)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#stereo-seq-stereoseq-stereoseq_nosplit" class="md-nav__link">
    <span class="md-ellipsis">
      
        Stereo-seq (stereoseq, stereoseq_nosplit)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#custom-platform-custom_sc" class="md-nav__link">
    <span class="md-ellipsis">
      
        Custom platform (custom_sc)
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              
              <article class="md-content__inner md-typeset">
                
                  



<h1 id="barcode-calling">Barcode calling</h1>
<p>IsoQuant includes a built-in barcode calling module (<code>isoquant_detect_barcodes.py</code>) that extracts
 barcodes and UMIs from raw long reads. This module supports multiple single-cell and
spatial transcriptomics platforms and can also be used as a standalone tool.</p>
<p>See <a href="single_cell.html">single-cell and spatial transcriptomics</a> for supported platforms,
pipeline integration, MDF format, and UMI deduplication.</p>
<h2 id="how-it-works">How it works</h2>
<p>The barcode calling module processes each read by:</p>
<ol>
<li>Searching for known constant sequences (linkers, primers, TSO) on the read</li>
<li>Locating the polyT tail to determine read orientation</li>
<li>Extracting the barcode region based on its expected position relative to the anchoring constant sequences</li>
<li>Matching the extracted barcode against a whitelist</li>
<li>Extracting the UMI sequence from its expected position</li>
</ol>
<p>Each read is assigned:
- A <strong>barcode</strong> (cell/spot identity) corrected against the whitelist, or <code>*</code> if no match is found
- A <strong>UMI</strong> (unique molecular identifier), or <code>*</code> if not detected
- A <strong>strand</strong> orientation (<code>+</code>, <code>-</code>, or <code>.</code> if unknown)
- Platform-specific features (polyT position, linker positions, etc.)</p>
<h2 id="integration-with-isoquant">Integration with IsoQuant</h2>
<p>When running IsoQuant with any single-cell or spatial mode, 
barcode calling is performed automatically as the first pipeline step.
The barcode detection output is a TSV file with one line per read, which is then used
throughout the rest of the pipeline for grouping and UMI deduplication.</p>
<p>If barcodes have already been called externally (e.g., by Cell Ranger),
use <code>--barcoded_reads</code> to skip the barcode calling step.</p>
<h2 id="standalone-usage">Standalone usage</h2>
<p><code>isoquant_detect_barcodes.py</code> can be run independently:</p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-0-1"><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a>python<span class="w"> </span>isoquant_detect_barcodes.py<span class="w"> </span><span class="se">\</span>
</span><span id="__span-0-2"><a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a><span class="w">  </span>--input<span class="w"> </span>reads.fastq.gz<span class="w"> </span><span class="se">\</span>
</span><span id="__span-0-3"><a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="w">  </span>--barcodes<span class="w"> </span>barcode_whitelist.txt<span class="w"> </span><span class="se">\</span>
</span><span id="__span-0-4"><a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="w">  </span>--mode<span class="w"> </span>tenX_v3<span class="w"> </span><span class="se">\</span>
</span><span id="__span-0-5"><a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="w">  </span>--output<span class="w"> </span>output_prefix<span class="w"> </span><span class="se">\</span>
</span><span id="__span-0-6"><a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="w">  </span>--threads<span class="w"> </span><span class="m">16</span>
</span></code></pre></div>
<h3 id="command-line-options">Command line options</h3>
<p><code>--input</code> or <code>-i</code> (required)</p>
<p>One or more input read files in FASTQ, FASTA or BAM format. Gzipped FASTQ/FASTA files are supported.</p>
<p><code>--output</code> or <code>-o</code> (required)</p>
<p>Output prefix. The barcode calling results will be written to <code>&lt;prefix&gt;.barcoded_reads.tsv</code>.
When multiple input files are provided, outputs are numbered: <code>&lt;prefix&gt;_0.barcoded_reads.tsv</code>, etc.</p>
<p><code>--barcodes</code> or <code>-b</code></p>
<p>One or more barcode whitelist files. The number of files depends on the mode:</p>
<ul>
<li>1 file: <code>tenX_v3</code>, <code>visium_5prime</code>, <code>stereoseq</code>, <code>stereoseq_nosplit</code>, <code>curio</code></li>
<li>2 files: <code>visium_hd</code></li>
<li>Not needed for <code>custom_sc</code> (barcodes defined in MDF file)</li>
</ul>
<p>File should contain one barcode sequence per line. 
More than 1 tab-separated column is allowed, but only the first will be used.
Supports plain text and gzipped files.</p>
<p><em>Note: barcode calling is performed much better if the whitelist contains a small number of barcodes. 
If you have a subset of barcodes from short-read data, provide them instead of the full whitelist.</em></p>
<p><code>--mode</code></p>
<p>Barcode calling mode. Available modes: <code>tenX_v3</code>, <code>curio</code>, <code>stereoseq</code>, <code>stereoseq_nosplit</code>,
<code>visium_5prime</code>, <code>visium_hd</code>, <code>custom_sc</code>. </p>
<p><code>--molecule</code>
Path to a molecule description format (MDF) file for <code>custom_sc</code> mode.
This file defines the structure of the sequencing molecule (barcodes, UMIs, linkers, polyT, cDNA)ÑŽ
See <a href="single_cell.html#molecule-definition-file-mdf-format">MDF format</a> for the format specification.</p>
<p><code>--threads</code> or <code>-t</code></p>
<p>Number of threads for parallel processing (default: 16).</p>
<p><code>--tmp_dir</code></p>
<p>Folder for temporary files during parallel processing.</p>
<h3 id="example-commands">Example commands</h3>
<p>10x Genomics:
<div class="language-bash highlight"><pre><span></span><code><span id="__span-1-1"><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a>python<span class="w"> </span>isoquant_detect_barcodes.py<span class="w"> </span><span class="se">\</span>
</span><span id="__span-1-2"><a id="__codelineno-1-2" name="__codelineno-1-2" href="#__codelineno-1-2"></a><span class="w">  </span>-i<span class="w"> </span>reads.fastq.gz<span class="w"> </span><span class="se">\</span>
</span><span id="__span-1-3"><a id="__codelineno-1-3" name="__codelineno-1-3" href="#__codelineno-1-3"></a><span class="w">  </span>-b<span class="w"> </span>3M-february-2018.txt.gz<span class="w"> </span><span class="se">\</span>
</span><span id="__span-1-4"><a id="__codelineno-1-4" name="__codelineno-1-4" href="#__codelineno-1-4"></a><span class="w">  </span>--mode<span class="w"> </span>tenX_v3<span class="w"> </span><span class="se">\</span>
</span><span id="__span-1-5"><a id="__codelineno-1-5" name="__codelineno-1-5" href="#__codelineno-1-5"></a><span class="w">  </span>-o<span class="w"> </span>results/my_sample<span class="w"> </span><span class="se">\</span>
</span><span id="__span-1-6"><a id="__codelineno-1-6" name="__codelineno-1-6" href="#__codelineno-1-6"></a><span class="w">  </span>-t<span class="w"> </span><span class="m">16</span>
</span></code></pre></div></p>
<p>Stereo-seq:
<div class="language-bash highlight"><pre><span></span><code><span id="__span-2-1"><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a>python<span class="w"> </span>isoquant_detect_barcodes.py<span class="w"> </span><span class="se">\</span>
</span><span id="__span-2-2"><a id="__codelineno-2-2" name="__codelineno-2-2" href="#__codelineno-2-2"></a><span class="w">  </span>-i<span class="w"> </span>reads.fastq.gz<span class="w"> </span><span class="se">\</span>
</span><span id="__span-2-3"><a id="__codelineno-2-3" name="__codelineno-2-3" href="#__codelineno-2-3"></a><span class="w">  </span>-b<span class="w"> </span>stereo_barcodes.txt<span class="w"> </span><span class="se">\</span>
</span><span id="__span-2-4"><a id="__codelineno-2-4" name="__codelineno-2-4" href="#__codelineno-2-4"></a><span class="w">  </span>--mode<span class="w"> </span>stereoseq<span class="w"> </span><span class="se">\</span>
</span><span id="__span-2-5"><a id="__codelineno-2-5" name="__codelineno-2-5" href="#__codelineno-2-5"></a><span class="w">  </span>-o<span class="w"> </span>results/stereo_sample<span class="w"> </span><span class="se">\</span>
</span><span id="__span-2-6"><a id="__codelineno-2-6" name="__codelineno-2-6" href="#__codelineno-2-6"></a><span class="w">  </span>-t<span class="w"> </span><span class="m">16</span>
</span></code></pre></div></p>
<p>Custom molecule:
<div class="language-bash highlight"><pre><span></span><code><span id="__span-3-1"><a id="__codelineno-3-1" name="__codelineno-3-1" href="#__codelineno-3-1"></a>python<span class="w"> </span>isoquant_detect_barcodes.py<span class="w"> </span><span class="se">\</span>
</span><span id="__span-3-2"><a id="__codelineno-3-2" name="__codelineno-3-2" href="#__codelineno-3-2"></a><span class="w">  </span>-i<span class="w"> </span>reads.fastq.gz<span class="w"> </span><span class="se">\</span>
</span><span id="__span-3-3"><a id="__codelineno-3-3" name="__codelineno-3-3" href="#__codelineno-3-3"></a><span class="w">  </span>--mode<span class="w"> </span>custom_sc<span class="w"> </span><span class="se">\</span>
</span><span id="__span-3-4"><a id="__codelineno-3-4" name="__codelineno-3-4" href="#__codelineno-3-4"></a><span class="w">  </span>--molecule<span class="w"> </span>my_platform.mdf<span class="w"> </span><span class="se">\</span>
</span><span id="__span-3-5"><a id="__codelineno-3-5" name="__codelineno-3-5" href="#__codelineno-3-5"></a><span class="w">  </span>-o<span class="w"> </span>results/custom_sample
</span></code></pre></div></p>
<h2 id="output">Output</h2>
<p>The main output is a TSV file (<code>*.barcoded_reads.tsv</code>) with a header line followed by one line per read.
The columns depend on the platform mode.</p>
<h3 id="common-columns-all-modes">Common columns (all modes)</h3>
<table>
<thead>
<tr>
<th>Column</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>read_id</td>
<td>Read identifier</td>
</tr>
<tr>
<td>barcode</td>
<td>Corrected barcode sequence, or <code>*</code> if not found</td>
</tr>
<tr>
<td>UMI</td>
<td>UMI sequence, or <code>*</code> if not detected</td>
</tr>
<tr>
<td>BC_score</td>
<td>Barcode alignment score (-1 if not matched)</td>
</tr>
<tr>
<td>strand</td>
<td>Read orientation: <code>+</code>, <code>-</code>, or <code>.</code></td>
</tr>
</tbody>
</table>
<h3 id="additional-columns-by-mode">Additional columns by mode</h3>
<p><strong>10x Genomics</strong> (<code>tenX_v3</code>, <code>visium_5prime</code>, <code>visium_hd</code>):</p>
<table>
<thead>
<tr>
<th>Column</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>polyT</td>
<td>Position of polyT tail start (-1 if not found)</td>
</tr>
<tr>
<td>R1</td>
<td>Position of R1 primer end (-1 if not found)</td>
</tr>
</tbody>
</table>
<p><strong>Curio</strong> (<code>curio</code>):</p>
<table>
<thead>
<tr>
<th>Column</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>polyT</td>
<td>PolyT tail position</td>
</tr>
<tr>
<td>primer</td>
<td>PCR primer position</td>
</tr>
<tr>
<td>linker_start</td>
<td>Linker start position</td>
</tr>
<tr>
<td>linker_end</td>
<td>Linker end position</td>
</tr>
</tbody>
</table>
<p><strong>Stereo-seq</strong> (<code>stereoseq_nosplit</code>, <code>stereoseq</code>):</p>
<table>
<thead>
<tr>
<th>Column</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>polyT</td>
<td>PolyT tail position</td>
</tr>
<tr>
<td>primer</td>
<td>Primer position</td>
</tr>
<tr>
<td>linker_start</td>
<td>Linker start position</td>
</tr>
<tr>
<td>linker_end</td>
<td>Linker end position</td>
</tr>
<tr>
<td>TSO</td>
<td>TSO position (-1 if not found)</td>
</tr>
</tbody>
</table>
<p><strong>Custom</strong> (<code>custom_sc</code>): columns depend on the elements defined in the MDF file.
Each variable element produces a column with its detected sequence and score.</p>
<h3 id="statistics-file">Statistics file</h3>
<p>A statistics file (<code>*.barcoded_reads.tsv.stats</code>) is generated alongside the TSV output,
reporting the total number of reads processed, barcodes detected, and UMIs found.</p>
<h2 id="supported-platforms">Supported platforms</h2>
<h3 id="10x-genomics-tenx_v3-visium_5prime">10x Genomics (<code>tenX_v3</code>, <code>visium_5prime</code>)</h3>
<p>Molecule structure (3' to 5'):
<div class="language-text highlight"><pre><span></span><code><span id="__span-4-1"><a id="__codelineno-4-1" name="__codelineno-4-1" href="#__codelineno-4-1"></a>[R1 primer] [Barcode (16bp)] [UMI (12bp)] [PolyT] [cDNA] [TSO]
</span></code></pre></div></p>
<p>Requires 1 barcode whitelist file (e.g., the 10x <code>3M-february-2018.txt.gz</code>).</p>
<h3 id="visium-hd-visium_hd">Visium HD (<code>visium_hd</code>)</h3>
<p>Molecule structure (3' to 5'):
<div class="language-text highlight"><pre><span></span><code><span id="__span-5-1"><a id="__codelineno-5-1" name="__codelineno-5-1" href="#__codelineno-5-1"></a>[R1 primer] [UMI (9bp)] [BC_part1 (15-16bp)] [BC_part2 (14-15bp)] VV [PolyT] [cDNA] [TSO]
</span></code></pre></div></p>
<p>Requires 2 barcode whitelist files (part 1 and part 2).
The final barcode is the concatenation of both parts.</p>
<h3 id="curio-bioscience-curio">Curio Bioscience (<code>curio</code>)</h3>
<p>Molecule structure (3' to 5'):
<div class="language-text highlight"><pre><span></span><code><span id="__span-6-1"><a id="__codelineno-6-1" name="__codelineno-6-1" href="#__codelineno-6-1"></a>[PCR primer] [Left BC (8bp)] [Linker] [Right BC (6bp)] [UMI (9bp)] [PolyT] [cDNA]
</span></code></pre></div></p>
<p>The linker sequence anchors the barcode detection.
Accepts 1 combined whitelist file (14bp barcodes).</p>
<h3 id="stereo-seq-stereoseq-stereoseq_nosplit">Stereo-seq (<code>stereoseq</code>, <code>stereoseq_nosplit</code>)</h3>
<p>Molecule structure (3' to 5'):
<div class="language-text highlight"><pre><span></span><code><span id="__span-7-1"><a id="__codelineno-7-1" name="__codelineno-7-1" href="#__codelineno-7-1"></a>[Primer] [Barcode (25bp)] [Linker] [UMI (10bp)] [PolyT] [cDNA] [TSO]
</span></code></pre></div></p>
<ul>
<li><code>stereoseq</code> mode: splits concatenated reads at TSO boundaries, producing a new FASTA with individual subreads</li>
<li><code>stereoseq_nosplit</code> mode: processes reads without splitting</li>
</ul>
<h3 id="custom-platform-custom_sc">Custom platform (<code>custom_sc</code>)</h3>
<p>Uses a molecule definition file (MDF) to describe arbitrary molecule structures.
See <a href="single_cell.html#molecule-definition-file-mdf-format">MDF format</a>.</p>
<p>The universal extractor:
1. Tries both forward and reverse complement orientations
2. Detects polyT tail to determine strand
3. Finds constant elements (primers, linkers) on the read
4. Extracts variable elements (barcodes, UMIs) from their expected positions
5. Corrects barcodes against whitelists when provided (<code>VAR_FILE</code> or <code>VAR_LIST</code>)
6. Handles linked elements: concatenated (<code>|</code>) parts are joined then corrected; duplicated (<code>/</code>) parts use majority vote</p>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      
      <script id="__config" type="application/json">{"annotate": null, "base": ".", "features": ["content.code.copy"], "search": "assets/javascripts/workers/search.2c215733.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="assets/javascripts/bundle.79ae519e.min.js"></script>
      
    
  </body>
</html>