<html>
<title>IsoQuant changelog</title>
<body>

<h3>IsoQuant 3.4.0, 3 May 2024</h3>

<p>Major novelties and improvements:</p>

<p>Significant speed-up on datasets containing regions with extremely high coverage,
    often encountered on mitochondrial chromosomes (<a href="https://github.com/ablab/IsoQuant/issues/97">#97</a>).</p>

<p>Added support for Illumina reads for spliced alignment correction (thanks to @rkpfeil).</p>

<p>Added support YAML files (thanks to @rkpfeil).</p>

<p>Transcript discovery and GTF processing:</p>

<p>Fixed missing genes in extended GTF (<a href="https://github.com/ablab/IsoQuant/issues/140">#140</a>,
    <a href="https://github.com/ablab/IsoQuant/issues/147">#147</a>, <a href="https://github.com/ablab/IsoQuant/issues/151">#151</a>,
    <a href="https://github.com/ablab/IsoQuant/issues/175">#175</a>).</p>

<p>Fixed strand detection and output of transcripts with <code>.</code> strand (<a href="https://github.com/ablab/IsoQuant/issues/107">#107</a>).</p>

<p>Added <code>--report_canonical</code> and <code>--polya_requirement</code> options that allows to control level of
    filtering of output transcripts based on canonical splice sites and the presence of poly-A tails. (<a href="https://github.com/ablab/IsoQuant/issues/128">#128</a>)</p>

<p>Added check for input GTFs (<a href="https://github.com/ablab/IsoQuant/issues/155">#155</a>).</p>

<p>Extract CDS and other features from reference GTF to the output GTFs (<a href="https://github.com/ablab/IsoQuant/issues/176">#176</a>).</p>

<p>Improved novel gene naming (<a href="https://github.com/ablab/IsoQuant/issues/164">#164</a>) </p>

<p>Read assignment and quantification:</p>

<p>Optimized read assignment algorithm.</p>

<p>Added <code>gene_assignment_type</code> attribute to read assignments.</p>

<p>Fixed duplicated records in <code>read_assignments.tsv</code> (<a href="https://github.com/ablab/IsoQuant/issues/168">#168</a>).</p>

<p>Improved reference gene and transcript quantification. Only unique assignments are now used for transcript quantification.</p>

<p>Improved consistency between <code>trascript_counts.tsv</code> and <code>transcript_model_counts.tsv</code>
    (<a href="https://github.com/ablab/IsoQuant/issues/127">#127</a>).</p>

<p>Introduced mapping quality filtering: <code>--min_mapq</code>, <code>--inconsistent_mapq_cutoff</code> and
    <code>--simple_alignments_mapq_cutoff</code> (<a href="https://github.com/ablab/IsoQuant/issues/110">#110</a>).</p>

<p>Minor fixes and improvements:</p>

<p>Added <code>--bam_tags</code> option to import additional information from BAM files to read assignments output.</p>

<p>Large output files are now gzipped by default, <code>--no_gzip</code> can be used to keep uncompressed output
    (<a href="https://github.com/ablab/IsoQuant/issues/154">#154</a>).</p>

<p>BAM stats are now printed to the log (<a href="https://github.com/ablab/IsoQuant/issues/139">#139</a>).</p>

<p>Various minor fixes and requests (<a href="https://github.com/ablab/IsoQuant/issues/106">#106</a>,
    <a href="https://github.com/ablab/IsoQuant/issues/141">#141</a>, <a href="https://github.com/ablab/IsoQuant/issues/143">#143</a>,
    <a href="https://github.com/ablab/IsoQuant/issues/146">#146</a>)</p>


<h3>IsoQuant 3.3.1, 26 July 2023</h3>

<p>Fixed "<code>UnboundLocalError: local variable 'match' referenced before assignment</code>" error in SQANTI-like output.</p>

<h3>IsoQuant 3.3, 13 June 2023</h3>

<p>Fixed read to novel models assignment.</p>

<p>Improved command line options for providing multiple files, added <code>--prefix</code> option.</p>

<p>Additional checks for various unusual cases in input GTFs.</p>

<p>Do not output empty files when no GTF is provided.</p>

<h3>IsoQuant 3.2, 27 March 2023</h3>

<p>Unspliced novel transcripts are not reported by the default for ONT data, use <code>--report_novel_unspliced</code> to generate them.</p>

<p>When multiple BAM/FASTQ files are provided via <code>--bam</code> / <code>--fastq</code>,
    they are treated as different replicas/samples of the same experiment;
    a single GTF and per-sample counts are generated automatically.</p>

<p>10-15 times lower RAM consumption with the same running time.</p>

<p>~5 times lower disk consumption for temporary files.</p>

<p><code>--low_memory</code> option has no effect (used by default); <code>--high_memory</code> mimics old behavior by storing alignments in RAM.</p>

<p>Read assignment reports transcript start and end (TSS/TES) matches.</p>

<p><code>--sqanti_output</code> generates SQANTI-like output for novel vs reference transcripts.</p>

<p>Resulting annotation contains exon ids.</p>

<p>Supplementary gene attributes are copied from the reference annotation to the output annotations.</p>

<p>Improved <code>--resume</code> and <code>--force</code> behaviour.</p>

<p><code>--model_construction_strategy sensitive_pacbio</code> is now more sensitive.</p>

<h3>IsoQuant 3.1.2, 7 February 2023</h3>

<p>Fixed strand detection that caused lower precision for novel transcripts.</p>

<p>Fixed known transcript filtering that caused lower recall.</p>

<p>Fixed duplicate transcript entries in the output annotation.</p>

<p>Fixed duplicate canonical attribute in extended annotation.</p>

<p>Fix <code>--resume</code> option when relative paths were provided.</p>

<h3>IsoQuant 3.1.1, 16 January 2023</h3>

<p>Fixed error caused by introns of length 0 (strange corner case, but it does happen).</p>

<p>Fixed error when using a read grouping file.</p>

<h3>IsoQuant 3.1.0, 3 January 2023</h3>

<p>Implement <code>--resume</code> option for resuming failed runs.</p>

<p>Fix SQANTI-like output for raw reads.</p>

<p>Fix read strand detection, improves transcript discovery as well.</p>

<h3>IsoQuant 3.0.3, 17 October 2022</h3>

<p>Simplify transcript naming, IDs of known transcripts are preserved in the output.</p>

<p>More information about novel transcripts in GTF</p>

<h3>IsoQuant 3.0.2, 10 October 2022</h3>

<p>Fix GTF attributes, thanks to @rsalz.</p>

<h3>IsoQuant 3.0.1, 31 August 2022</h3>

<p>Fix <code>--check_canonical</code> option.</p>

<h3>IsoQuant 3.0.0, 31 August 2022</h3>

<p>Annotation-free mode for de novo transcript discovery.</p>

<p>Significant speed-up.</p>

<p>Extended annotation (all reference + novel transcripts) is now part of the output.</p>

<p>Intermediate BAM files have nicer names.</p>

<p>Proper single-thread mode without thread pool usage.

<h3>IsoQuant 2.3.0, 27 May 2022</h3>

<p>New options for controlling quantification strategies. Default behaviour is changed as well. </p>

<p>New option <code>--genedb_output</code> for providing a separate folder for gene database in case the
    output directory is located on a shared disk.</p>

<p>Possibility to provide read group tables in gzipped format.</p>

<h3>IsoQuant 2.2.2, 18 May 2022</h3>

<p>Fixed <code>--check_canonical</code> option.</p>

<p>Improved running time for the read assignment step (noticeable only for genes with > 100 exons).</p>

<h3>IsoQuant 2.2.1, 28 Apr 2022</h3>

<p>Minor fixes and improvement in output files. Note, that GTFs and some other files have now multiline headers.</p>

<h3>IsoQuant 2.2.0, 5 Mar 2022</h3>

<p>Parallel processing of transcript model construction phase.</p>

<p>Minor improvements in quantification of reference transcripts.</p>

<h3>IsoQuant 2.1.1, 7 Feb 2022</h3>

<p>Fixed counts/TPM for novel transcript models.</p>

<p>Fixed processing of BAM records without sequence data (e.g. secondary alignment).</p>

<p>Fixed <code>list index out of range</code> bug in long read counter.</p>

<h3>IsoQuant 2.1.0, 19 Jan 2022</h3>

<p>Improved recall by introducing relative coverage cutoffs.</p>

<p>More careful handling of transcript terminal positions.</p>

<p>Fixed GTF to BED conversion.</p>

<h3>IsoQuant 2.0.0, 7 Oct 2021</h3>

<p>Completely new transcript discovery algorithm with significantly higher recall.</p>

<p>Algorithm for read alignment correction.</p>

<p>Support for technical replicas within a single sample.</p>

<h3>IsoQuant 1.3.0, 19 Jun 2021</h3>

<p>Significantly improved running time and RAM consumption;</p>

<p>Annotation is now fed into minimap2;</p>

<p>Extended output format.</p>

<h3>IsoQuant 1.2.2, 21 May 2021</h3>

<p>Support for GFF3 mRNA features.</p> 

<h3>IsoQuant 1.2.1, 21 Apr 2021</h3>

<p>Support for BAM files with =/X in CIGAR strings;</p> 

<p>Fixed canonical splice site detection.</p>

<h3>IsoQuant 1.2.0, 28 Mar 2021</h3>

<p>Multi-threading;</p>

<p>Intermediate results are saved to disc to enable quick restart via --read_assignments option;</p>

<p>Significantly improved precision for novel transcript detection;</p>

<p>Secondary alignments are now used by default;</p>

<p>Fixed several bugs in inconsistency detection algorithm;</p>

<p>Reworked polyA detection and reporting once again;</p>

<p>Slightly modified read assignment output format;</p>

<p>More informative GTF output;</p>

<p>Removed --has_polya option, --polya_trimmed is now used as the opposite;</p>

<p>Added --check_canonical option.</p>

<h3>IsoQuant 1.1.0, 11 Dec 2020</h3>

<p>Significantly reworked polyA detection and reporting;</p>

<p>Improved detection of inconsistencies, added several new event types;</p>

<p>Better recall and precision for read assignment algorithm;</p>

<p>Fixed several bug and flaws;</p>

<p>Added script for counting simple stats for GTF files (srt/gtf_stats.py).</p>

<h3>IsoQuant 1.0.0, 12 Jul 2020</h3>

<p>Initial release.</p>

<p>
    <a href="https://github.com/ablab/IsoQuant">IsoQuant GitHub page</a>
</p>

</body>
</html>
